uniffi_core/src/ffi/rustfuture/mod.rs

```diff
@@ -30,6 +30,32 @@ pub enum RustFuturePoll {
 /// to continue progress on the future.
 pub type RustFutureContinuationCallback = extern "C" fn(callback_data: u64, RustFuturePoll);
 
+/// This marker trait allows us to put different bounds on the `Future`s we support, based on
+/// `#[cfg()]` configuration.
+pub trait UniffiCompatibleFuture<T, E>: Future<Output = Result<T, E>> {}
+
+#[cfg(not(target_arch = "wasm32"))]
+impl<T, F, E> UniffiCompatibleFuture<T, E> for F where F: Future<Output = Result<T, E>> + Send {}
+
+/// # Safety:
+///
+/// WASM32 is a single threaded environment. However, in a browser there do exist
+/// [`WebWorker`](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers)s which do
+/// not share memory or event-loop with the main browser context.
+///
+/// The most common source of asynchrony,
+/// [`JsPromise`](https://github.com/rustwasm/wasm-bindgen/blob/main/crates/futures/src/lib.rs)
+/// is not `Send` because:
```

Comment:
This seems like the most important point here.  We're just following the rules that wasm-bindgen already set, I'd love that to be emphasized.  Maybe something like "UniFFI is following the path of the most common source of asynchrony, ..."